File: apps/write-8-minutes/ios/Anky/Purchases/AnkyPurchasesConfig.swift
import Foundation
import RevenueCat

/// The one configuration surface for everything purchases. RevenueCat is
/// the single transaction owner (it observes and finishes every StoreKit
/// transaction); the app never talks to StoreKit directly.
///
/// These values must stay in sync with the RevenueCat dashboard
/// (entitlement + offering), Anky.storekit (local testing), and App Store
/// Connect (production).
enum AnkyPurchasesConfig {
    /// RevenueCat public Apple API key (safe to ship in the binary).
    static let revenueCatPublicKey = "appl_mvCsxolPWZmQjtULGLQhmOUhGMY"

    /// The single entitlement every gate checks — purchased or granted.
    static let entitlementID = "pro"

    /// The paywall's offering; its packages carry yearly and monthly.
    static let offeringID = "default"

    static let subscriptionGroupName = "Anky"
    static let yearlyProductID = "anky.yearly"
    static let monthlyProductID = "anky.monthly"
    static let allProductIDs: Set<String> = [yearlyProductID, monthlyProductID]

    /// The honest reminder fires this long BEFORE the trial converts —
    /// "your trial ends tomorrow", while there is still real time to
    /// cancel. (28h before the end of a 72h trial ≈ the old 44h-in mark.)
    static let trialReminderLeadTime: TimeInterval = 28 * 60 * 60
}

/// Launch-time RevenueCat configuration. Called from AppRoot once identity
/// is loaded — never lazily from a feature path, so every surface sees the
/// same configured SDK from the first frame on.
///
/// Pattern choice: `Purchases.configure(withAPIKey:appUserID:)` with the
/// wallet address directly, NOT configure-anonymous-then-logIn. Identity
/// is created synchronously before this runs, so there is never a window
/// where a purchase could attach to an anonymous RevenueCat ID. If
/// identity creation fails we fail closed: the SDK stays unconfigured, no
/// purchase path opens, and the next foreground retries.
enum AnkyPurchases {
    @MainActor
    private(set) static var configuredAccountId: String?

    /// True once the SDK is configured under a real writer identity.
    @MainActor
    static var isConfigured: Bool {
        configuredAccountId != nil
    }

    /// Configures on first call, re-identifies via `logIn` if the wallet
    /// changed (iCloud restore adopting another writer's identity).
    /// Returns the active appUserID, or nil when identity is unavailable
    /// (fail closed — retry on next foreground).
    @MainActor
    @discardableResult
    static func identifyCurrentWriter(
        identityStore: WriterIdentityStore = WriterIdentityStore()
    ) async -> String? {
        guard let identity = try? identityStore.loadOrCreate(),
              !identity.address.isEmpty else {
            return nil
        }
        let accountId = identity.address

        guard let configured = configuredAccountId else {
            #if DEBUG
            Purchases.logLevel = .debug
            #endif
            Purchases.configure(
                withAPIKey: AnkyPurchasesConfig.revenueCatPublicKey,
                appUserID: accountId
            )
            configuredAccountId = accountId
            return accountId
        }

        guard configured != accountId else {
            return accountId
        }
        guard (try? await Purchases.shared.logIn(accountId)) != nil else {
            return configured
        }
        configuredAccountId = accountId
        return accountId
    }
}

File: apps/android/app/src/main/java/inc/anky/android/core/subscription/AnkyPurchasesConfig.kt
package inc.anky.android.core.subscription

import android.content.Context
import com.revenuecat.purchases.LogLevel
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.awaitLogIn
import inc.anky.android.BuildConfig
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * The one configuration surface for everything purchases. RevenueCat is the
 * single transaction owner (it observes and finishes every Play Billing
 * transaction); the app never talks to Play Billing directly.
 *
 * These values must stay in sync with the RevenueCat dashboard (entitlement +
 * offering) and Play Console (production). The product ids intentionally match
 * iOS (`anky.yearly` / `anky.monthly`): Play Console must define subscriptions
 * with these exact ids so RevenueCat's cross-platform `pro` entitlement maps
 * one-to-one. On Android the yearly base plan carries the 3-day free trial as
 * an offer; Play Billing only returns trial offers the writer is eligible for.
 *
 * Ported from iOS `Anky/Purchases/AnkyPurchasesConfig.swift`.
 */
object AnkyPurchasesConfig {
    /** The single entitlement every gate checks — purchased or granted. */
    const val ENTITLEMENT_ID = "pro"

    /** The paywall's offering; its packages carry yearly and monthly. */
    const val OFFERING_ID = "default"

    const val YEARLY_PRODUCT_ID = "anky.yearly"
    const val MONTHLY_PRODUCT_ID = "anky.monthly"
    val ALL_PRODUCT_IDS: Set<String> = setOf(YEARLY_PRODUCT_ID, MONTHLY_PRODUCT_ID)

    /**
     * The honest reminder fires this long BEFORE the trial converts —
     * "your trial ends tomorrow", while there is still real time to cancel.
     * (28h before the end of a 72h trial ≈ the old 44h-in mark.)
     */
    const val TRIAL_REMINDER_LEAD_TIME_MILLIS: Long = 28L * 60L * 60L * 1000L
}

/**
 * Launch-time RevenueCat configuration. Called from the app root once identity
 * is loaded — never lazily from a feature path, so every surface sees the same
 * configured SDK from the first frame on.
 *
 * Pattern choice: `Purchases.configure` with the wallet address as appUserID
 * directly, NOT configure-anonymous-then-logIn. Identity is created
 * synchronously before this runs, so there is never a window where a purchase
 * could attach to an anonymous RevenueCat ID. If identity is unavailable we
 * fail closed: the SDK stays unconfigured, no purchase path opens, and the
 * next foreground retries.
 *
 * Ported from iOS `AnkyPurchases` in `AnkyPurchasesConfig.swift`. The API key
 * comes from `BuildConfig.REVENUECAT_ANDROID_PUBLIC_KEY` (a public key, safe
 * to ship in the binary).
 */
object AnkyPurchases {
    private val mutex = Mutex()

    /** The appUserID the SDK is configured under, or null while unconfigured. */
    @Volatile
    var configuredAccountId: String? = null
        private set

    /** True once the SDK is configured under a real writer identity. */
    val isConfigured: Boolean
        get() = configuredAccountId != null

    /**
     * Configures on first call, re-identifies via `logIn` if the wallet
     * changed (a backup import adopting another writer's identity). Returns
     * the active appUserID, or null when identity is unavailable (fail
     * closed — retry on next foreground).
     *
     * @param appUserId the writer's wallet address — never an anonymous id.
     */
    suspend fun identifyCurrentWriter(context: Context, appUserId: String?): String? =
        mutex.withLock {
            val apiKey = BuildConfig.REVENUECAT_ANDROID_PUBLIC_KEY.trim()
            if (apiKey.isBlank() || appUserId.isNullOrBlank()) {
                return@withLock null
            }

            val configured = configuredAccountId
            if (configured == null) {
                Purchases.logLevel = if (BuildConfig.DEBUG) LogLevel.DEBUG else LogLevel.ERROR
                val outcome = runCatching {
                    if (!Purchases.isConfigured) {
                        Purchases.configure(
                            PurchasesConfiguration.Builder(context.applicationContext, apiKey)
                                .appUserID(appUserId)
                                .diagnosticsEnabled(false)
                                .automaticDeviceIdentifierCollectionEnabled(false)
                                .build(),
                        )
                    } else if (Purchases.sharedInstance.appUserID != appUserId) {
                        // Another surface (legacy credits client) configured
                        // first — adopt the shared instance under our writer.
                        Purchases.sharedInstance.awaitLogIn(appUserId)
                    }
                }
                if (outcome.isFailure) {
                    return@withLock null
                }
                configuredAccountId = appUserId
                return@withLock appUserId
            }

            if (configured == appUserId) {
                return@withLock appUserId
            }
            val logIn = runCatching { Purchases.sharedInstance.awaitLogIn(appUserId) }
            if (logIn.isFailure) {
                return@withLock configured
            }
            configuredAccountId = appUserId
            appUserId
        }
}
